src/main/java/com/google/songgame/servlets/RoomServlet.java

```diff
@@ -1,19 +1,55 @@
 package com.google.songgame.servlets;
 
+import com.google.gson.Gson;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.stream.Collectors;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.hc.core5.http.ParseException;
+import com.google.gson.Gson; 
 
 @WebServlet("/room")
 public final class RoomServlet extends HttpServlet {
 
+  private Gson gson;
+
+  @Override
+  public void init() {
+    gson = new Gson();
+  }
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get user name
-    String name = request.getParameter("user-name");
-    response.setContentType("text/html;");
-    response.getWriter().println("Name: " + name);
+    Map<String, String> userProperties = readJSONFromRequest(request);
+
+    // Save player username and userId to datastore.
+    Entity userEntity = new Entity("User");
+    userEntity.setProperty("username", userProperties.get("username"));
+    userEntity.setProperty("userId", userProperties.get("userId"));
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(userEntity);
+  }
+
+  private Map<String, String> readJSONFromRequest(HttpServletRequest request) throws IOException {
+    String requestJSONString = request.getReader().lines().collect(Collectors.joining());
+    Map<String, String> jsonData = gson.fromJson(requestJSONString, Map.class);
```

Comment:
I'm actually surprised this is working. However, I think it's safer to provide a TypeToken like we recommended Salil do in his code. Can you add the following constant to this class:
```
private final static Type MESSAGE_TYPE = new TypeToken<Map<String, String>>(){}.getType();
```

Like is done here:
https://github.com/googleinterns/step105-2020/blob/master/src/main/java/com/google/songgame/servlets/ChatServlet.java#L25

And then pass that constant as the second parameter to this method, like is done here:
https://github.com/googleinterns/step105-2020/blob/master/src/main/java/com/google/songgame/servlets/ChatServlet.java#L52

Reply:
@pruess-google Fixed